scipy/spatial/distance.py

```diff
@@ -130,6 +164,31 @@ def _copy_array_if_base_present(a):
     return a
 
 
+def _correlation_cdist_wrap(XA, XB, dm, **kwargs):
+    XA -= XA.mean(axis=1)[:, np.newaxis]
+    XB -= XB.mean(axis=1)[:, np.newaxis]
+    _cosine_cdist_wrap(XA, XB, dm, **kwargs)
+
+
+def _correlation_pdist_wrap(X, dm, **kwargs):
+    X2 = X - X.mean(1)[:, np.newaxis]
+    norms = _row_norms(X2)
+    _distance_wrap.pdist_cosine_wrap(X2, dm, norms, **kwargs)
+
+
+def _cosine_cdist_wrap(XA, XB, dm):
+    np.dot(XA, XB.T, out=dm)
+    dm /= _row_norms(XA).reshape(-1, 1)
+    dm /= _row_norms(XB)
+    dm *= -1
+    dm += 1
+
+
+def _cosine_pdist_wrap(X, dm, **kwargs):
+    norms = _row_norms(X)
+    _distance_wrap.pdist_cosine_wrap(X, dm, norms, **kwargs)
+
+
```

Comment:
These functions allow to treat cosine and correlation as other metrics in `_SIMPLE_XDIST` and further clean code